fix: PLT-863: Remove iterator usages

triklozoid opened the pull request:
Changes

  - Added iterate_queryset() function in core.utils.iterators for efficient iteration over large QuerySets using ID-based chunking
  - Replaced all .iterator() usages in production code with iterate_queryset()
  - Added QS_ITERATOR_DEFAULT_CHUNK_SIZE setting (env variable, default=1000)
  -  Two places where `iterator` was just removed without replacement was added by me, I confirm, that it's safe and will not cause any OOMs

  Reason

  - .iterator() doesn't work with pgBouncer in transaction pooling mode



<!--

This description MUST be filled out for a PR to receive a review. Its primary purposes are:

 - to enable your reviewer to review your code easily, and
 - to convince your reviewer that your code works as intended.

Some pointers to think about when filling out your PR description:
 - Reason for change: Description of problem and solution
 - Screenshots: All visible changes should include screenshots.
 - Rollout strategy: How will this code be rolled out? Feature flags / env var / other
 - Testing: Description of how this is being verified
 - Risks: Are there any known risks associated with this change, eg to security or performance?
 - Reviewer notes: Any info to help reviewers approve the PR
 - General notes: Any info to help onlookers understand the code, or callouts to significant portions.

You may use AI tools such as Copilot Actions to assist with writing your PR description (see https://docs.github.com/en/copilot/using-github-copilot/using-github-copilot-for-pull-requests/creating-a-pull-request-summary-with-github-copilot); however, an AI summary isn't enough by itself. You'll need to provide your reviewer with strong evidence that your code works as intended, which requires actually running the code and showing that it works.

-->

triklozoid committed:
fix: PLT-863: Remove iterator usages

triklozoid commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17234686311)
> Successfully merged:  7 files changed, 47 insertions(+), 17 deletions(-)

nikitabelonogov commented:
/fm sync
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17235998442)

nikitabelonogov committed:
Sync Follow Merge dependencies

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17235998442

nikitabelonogov committed:
Merge branch 'develop' into 'fb-PLT-863'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17235998442

matt-bernstein commented:
Could you add a cursorrule for this pattern? The examples @hakan458 posted in slack would be a good start for it https://humansignal.slack.com/archives/C03DQ8U76PP/p1756165907609189?thread_ts=1755891562.628529&cid=C03DQ8U76PP

triklozoid commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17250678787)
> Successfully merged:  create mode 100644 docs/source/guide/release_notes/onprem/2.26.1.md

triklozoid committed:
Merge branch 'develop' into 'fb-PLT-863'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17250678787

triklozoid committed:
Add cursorrule

triklozoid committed:
Add feature flag

makseq commented on label_studio/core/utils/iterators.py:
```diff
@@ -0,0 +1,33 @@
+
+        # Create a new queryset based on the original, preserving all optimizations:
+        # annotations, select_related, prefetch_related, only/defer
+        chunk_qs = queryset.filter(**{f'{pk_field}__in': chunk_ids})
```
This does not guarantee increasing ordering. Did iterator() do it?

triklozoid replied:
`iterator()` did it, but in my opinion most of the places does not require the order, we just need to process all objects from queryset. I'll add note about it in cursorrule.

triklozoid committed:
Update cursorrule

hakan458 commented on label_studio/core/utils/iterators.py:
```diff
@@ -0,0 +1,33 @@
+    if chunk_size <= 0:
+        raise ValueError(f'chunk_size must be positive, got {chunk_size}')
+
+    if not flag_set('fflag_fix_back_plt_863_remove_iterator_27082025_short', user='auto'):
```
do we need a FF for this? Why would we ever turn off the FF and use normal iterator? On prem I guess?

triklozoid replied:
I added it for the case, if a change in query patterns will cause some unexpected database load, to be able to quickly revert the change and avoid long downtime. It can be removed after release, if everything is fine.

triklozoid commented:
/git merge
> [Workflow run](https://github.com/HumanSignal/label-studio/actions/runs/17295731684)
> Successfully merged:  create mode 100644 web/libs/editor/src/utils/errors.js

triklozoid committed:
Merge branch 'develop' into 'fb-PLT-863'

Workflow run: https://github.com/HumanSignal/label-studio/actions/runs/17295731684
